misc/mdns/src/service.rs

```diff
@@ -147,33 +153,60 @@ impl MdnsService {
         Ok(MdnsService {
             socket,
             query_socket: UdpSocket::bind((Ipv4Addr::from([0u8, 0, 0, 0]), 0u16)).await?,
-            query_interval: Interval::new(Duration::from_secs(20)),
+            query_interval: Interval::new_at(Instant::now(), Duration::from_secs(20)),
             silent,
             recv_buffer: [0; 2048],
             send_buffers: Vec::new(),
             query_send_buffers: Vec::new(),
         })
     }
 
-    pub async fn next_packet(&mut self) -> MdnsPacket {
-        // TODO: refactor this block
+    pub fn enqueue_response(&mut self, rsp: Vec<u8>) {
+        self.send_buffers.push(rsp);
+    }
+
+    /// Returns a future resolving to itself and the next received `MdnsPacket`.
+    ///
+    /// **Note**: Why does `next` take ownership of itself?
+    ///
+    /// `MdnsService::next` needs to be called from within `NetworkBehaviour`
+    /// implementations. Given that traits can not have async methods the
+    /// respective `NetworkBehaviour` implementation needs to somehow keep the
+    /// Future returned by `MdnsService::next` across classic `poll`
+    /// invocations. The instance method `next` can either take a reference or
+    /// ownership of itself:
+    ///
+    /// 1. Taking a reference - If `MdnsService::poll` takes a reference to
+    /// `&self` the respective `NetworkBehaviour` implementation would need to
+    /// keep both the Future as well as its `MdnsService` instance across poll
+    /// invocations. Given that in this case the Future would have a reference
+    /// to `MdnsService`, the `NetworkBehaviour` implementation struct would
+    /// need to be self-referential which is not possible without unsafe code in
+    /// Rust.
+    ///
+    /// 2. Taking ownership - Instead `MdnsService::next` takes ownership of
+    /// self and returns it alongside an `MdnsPacket` once the actual future
+    /// resolves, not forcing self-referential structures on the caller.
+    pub async fn next(mut self) -> (Self, MdnsPacket) {
         // Send a query every time `query_interval` fires.
+        //
         // Note that we don't use a loop here—it is pretty unlikely that we need it, and there is
         // no point in sending multiple requests in a row.
-        match Stream::poll_next(Pin::new(&mut self.query_interval), cx) {
-            Poll::Ready(_) => {
-                if !self.silent {
-                    let query = dns::build_query();
-                    self.query_send_buffers.push(query.to_vec());
-                }
+        if let Some(_) = self.query_interval.next().now_or_never() {
```

Comment:
That's not correct, unfortunately.
Once you reach `recv().await`, you will block forever until the socket has data for you.
Instead, what we want is for `query_interval` to "interrupt" that and dispatch a new request.